### PR TITLE
[SPARK-6746B] Refactor large functions in DAGScheduler to improve readibility

### PR DESCRIPTION
**Pull request by ilganeli**
Hi all. The DAG Scheduler had several very large, complex, and unreadable functions. Specifically, 
`handleTaskCompletion`, `cleanupStateForJobAndIndependentStages`, and `submitMissingTasks`. This patch is almost exclusively code movement, very little code has changed with the exception of handling return logic when serializing tasks within `submitMissingTasks`. 

I believe the final product is much more readable and makes it much easier to understand what's going on within the DAGScheduler.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -418,6 +418,32 @@ class DAGScheduler(
     updateJobIdStageIdMapsList(List(stage))
   }
 
+  /** Clear references to the given stage in stored data structures */
+  private def removeStage(stageId: Int) {
```
Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: By moving this you've changed the scope in which removeStage is visible, potentially opening up improper calls to removeStage from elsewhere than intended in the DAGScheduler.

**Comment by markhamstra** [comment]
Ok, in looking further, I see that I'm going to have a more general problem with this PR.  I can't see exposing carefully nested and scoped portions of functions to the rest of the DAGScheduler as a good thing.  Things are defined and visible where they make sense and are needed, and are intentionally not available elsewhere.  To me, exposing additional interfaces to all of the DAGScheduler makes it more complicated to figure out how and when the various pieces should be used, not less.

**Comment by ilganeli** [comment]
Mark - your point is valid. However, then we we get enormous monolithic functions that are near impossible to read , comprehend, or debug without specialized knowledge. The counterpoint I'd make is that any changes throughout DAGScheduler are necessarily faced with heavy scrutiny, I'd argue that opportunity for misuse of the now visible functions is unlikely.

Moreover, because their functionality is now much clearer, it will be easier to reason about any subsequent changes and write cleaner more stable code that uses these pieces effectively.



Sent with Good (www.good.com)


-----Original Message-----

**Comment by markhamstra** [comment]
I guess I just don't see things the same way.  I don't have a problem with the DAGScheduler being implemented with scoping idioms that may be unfamiliar to a large number of coders.  The DAGScheduler is about as far from being a public API in Spark as you can get, and I don't really think we want to sacrifice the additional safety that nested scopes provide within the DAGScheduler just to make it easy for more developers to make changes to that part of Spark's code.  If you want an even more contrary point of view, a high bar of expected familiarity with certain idioms actually can serve us well in restricting just who is making changes to the DAGScheduler and what style and concerns they are maintaining.

Quite simply, I don't find a function declared within the scope of another function, for example, to be at all difficult to read -- it's just a common idiom across several programming languages with which I am familiar.  On the other hand, flattening out those carefully nested scopes does actually make it harder for me to read and reason about where and under what preconditions a previously-nested function can and should be used.

I'm not saying that I won't consider any refactoring of the DAGScheduler or the flattening of any scopes; but in order for me not to veto them, any such changes will require a lot more justification than simply that making them produces code that is more readable for a larger audience of developers.

**Comment by ilganeli** [comment]
Mark - you've convinced me. With that said, do you see any way to make the code more modular without sacrificing the protection from scoping? My primary intent with my refactoring was to break things out into discrete function calls. Would moving those internal to where they're used help?



Sent with Good (www.good.com)


-----Original Message-----

**Comment by markhamstra** [comment]
Maybe... convince me.

**Comment by ilganeli** [comment]
@markhamstra I've moved the helper functions into the scope of the functions that use them and added clarifying comments to separate the helper functions from the code that uses them. The major gain that I see here is that now process flow for the complex functions is much easier to understand, e.g. in ```handleTaskCompletion``` the operational logic is now condensed to a concise set of case statements and the handling of each case is abstracted to the individual functions. 

By modularizing the code, it's now easier to reason about the operation of the DAGScheduler. I agree with the need to minimize uninformed modification of this code but I think even core committers could benefit from not having to consider execution logic that is several hundred lines long.

**Comment by sryza** [comment]
@markhamstra @ilganeli I'm strongly opposed to any line of thinking that obfuscates code to make it less accessible to those who aren't core contributors.  If we want to tell people to be careful around DAGScheduler, we can document this and be careful about accepting patches.  Eventually (and in part already), the few people familiar with the DAGScheduler internals will have other obligations, and trying to insulate knowledge about it will hurt the long term success of the project.

**Comment by markhamstra** [comment]
@sryza It's not at all about obfuscation for the sake of restricting the pool of contributors.  It is about not compromising code safety for the sake of increasing the pool of contributors.  At least in my opinion, the number of contributors to the DAGScheduler is not the driving concern.

**Comment by ilganeli** [comment]
@markhamstra What do you think of the latest revision? This makes the code more modular and readable (imo) without sacrificing anything by way of code safety.

**Comment by ilganeli** [comment]
All - I assume the lack of response indicates that folks aren't interested in this patch. I'll close this unless I get some further comments.

**Pull request closed without merging**